---
title: "Exercise 8: differential expression with count data"
output: html_document
---

In this exercise, we will explore two popular pipelines for differential expression of RNA-seq data, (already) given the counts.  Details on the counting and/or of the transcript abundance estimation were covered in the last lecture from Hubert and more details will be given in Charlotte's RNA-seq session.  The dataset used here is the well-known 'pasilla' data, which compares the knockout of pasilla (a splicing factor) to a wild-type control.

In this exercise, some code will be given to get started.  But, there are good resources on how to go through these fairly standard analysis pipelines (links below).

First, we have the samples already organized into a table of metadata and this is used to set the count filenames.

```{r}
library("edgeR")
samples <- read.table("data/samples.txt", header=TRUE,
                      row.names=5, stringsAsFactors=FALSE)
samples
```

Here, we read in the 7 count files and consolidate it into a single table as opposed to 7 individual files -- the readDGE() function saves having to do this manually:

```{r}
counts <- edgeR::readDGE(samples$countfile, path = "data/")$counts
```

Here, we can trim the column names to get a 'nice' grouping variable to use in the design matrix:

```{r}
grp <- gsub("\\-.[0-9]*","",colnames(counts))
grp
```


Below is a pipeline using likelihood-ratio tests that is adapted from Quick Start section, early in the [edgeR user's guide](http://bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf)


```{r eval=FALSE}
group <- factor(grp)
y <- edgeR::DGEList(counts=counts, group=group) 
y <- edgeR::calcNormFactors(y, method = "TMM") # adjust for different sizes of libraries
design <- model.matrix(~group)

y <- edgeR::estimateDisp(y,design) # estimate common, trended and tagwise dispersion
#To perform likelihood ratio tests:
fit <- edgeR::glmFit(y,design)
lrt <- edgeR::glmLRT(fit,coef=2) # DGELRT-class
edgeR::topTags(lrt) 
```

This will lead you to a list of putative/detected differentially expressed genes.  Note that the code above is just an example and is not actually run in the script.

#### Exercise 1: Fix the code above to work on the count table that was already read in.  Add in some spot checks, including an MDS plot from plotMDS(), a look at the dispersion-mean plot using plotBCV() and a look at the overall M vs A plot using plotSmear().


``` {r plotMDS}
limma::plotMDS(lrt$fitted.values)
edgeR::plotBCV(y) # counts per million
# subpopulations with different variance can be seen, lowess regression
edgeR::plotSmear(lrt)

```

For more details, see Section 15.5 Differential expression of the [limma user's guide](http://www.bioconductor.org/packages/devel/bioc/vignettes/limma/inst/doc/usersguide.pdf).

#### Exercise 2: For the MDS plot, investigate what happens differently when you call plotMDS() on the DGEList object and when you call it on the normalized counts -- for normalized counts, see the cpm() function.  Explain why it is different.

As an alternative (to edgeR above), one can run through a standard voom pipeline, such as:

```{r eval=FALSE}
v <- voom(d$counts, design=model.matrix(~grp), plot = TRUE)
vf = lmFit(v, design = model.matrix(~grp))  # note similarity to limma
                                            # analysis performed earlier
vf = eBayes(vf)
```

#### Exercise 3. Again, the above code for voom is not run here.  Fix this code to work with the count dataset above and then do some ``spot checks'' of the genes called DE.  For example, make a plot of the plot of estimated fold changes for the two methods, compare the P-values of the two methods, or create a venn diagram of the called sets of DE genes at a set threshold.
