---
title: "SAOHTGTD_project/EX8"
author: "Martin Holub"
date: "November 5, 2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r load_data}
# data_path <- "C:\\Users\\marti\\git\\sta426\\project\\Mouse_Processed_GTEx_Data.DGE.log-UMI-Counts.txt"
# anno_path <- "C:\\Users\\marti\\git\\sta426\\project\\Mouse_Meta_Data_with_cluster.txt"

#mouse.umi <- read.table(data_path, sep = "\t", header = TRUE, nrows = 1000)[, 1:2000]
#mouse.anno <- read.table(anno_path, sep = "\t", header = TRUE, nrows = 1000)[2:2000, ]

#move gene names out of matrix
rownames(mouse.umi) <- mouse.umi[ , 1]
mouse.umi <- mouse.umi[, -1]

nonzero_read_counts <- apply(mouse.umi, 1, function(r) sum(r != 0 )) # rowSums(mouse.umi != 0)
```

## Preprocessing

There are 13313 cells(nuclei) and 17308 genes. 

The data has been already:
* aligned to mm10 mouse UCSC reference genome using STAR v2.4.0a, recording exonic regions
  * barcodes filtered for minimum number of transcripts and aggregated (1 edit tolerance)
  * gene counts aggregated on unique UMIs (1 substitution tolerance)
  * reads with low Phred filtered out, trimmed from nucleus barcode (12 bases) and unique molecular index (8 bases)
* transcript counts were assembled into digital gene expression (DGE) matrix
  * the DGE was scaled by total number of UMI counts, multiplied per by mean number of transcripts and log transformed
* effects of number of transcripts and genes detected per nucleus were [regressed out](http://satijalab.org/seurat/cell_cycle_vignette.html)
* Nuclei with less than 200 genes and less than 10000 suable reads were filtered out
* Genes detected in les then 10 nuclei were removed
* Highly variable genes vere selected by fitting gamma-distribution relationship between mean counts and coefficent of variation. Genes with difference in coeficient of variation between empirical and expected lower than 0.2 were dropped. Also genes with mean transcript count lower than 0.005 were dropped.

## Dimensionality reduction

Next, PCA is done and number of components to retain is selected by method based on [largest eigenvalue gap](http://science.sciencemag.org/content/sci/suppl/2016/07/27/science.aad7038.DC1/Habib.SM.pdf).

``` {r pca}
mouse.pca <- rsvd::rpca(mouse.umi, k = 5, center = FALSE, scale = FALSE) # just pick some k 

```

## Graph clustering

The use [custom script](https://github.com/broadinstitute/BipolarCell2016/blob/master/class.R). -.-

## Subclustering

At least the use the same algorithm here  -.-.


## tSNE

``` {r }
mouse.rtsne <- Rtsne::Rtsne(mouse.pca$rotation, pca = FALSE, perplexity = 100, max_iter = 2000)
plot(mouse.rtsne$Y, col = mouse.anno$ClusterID)
```

## We start once again and this time try to reproduce the steps that lead from raw to preprocessed data

``` {r paths}
data_dir <- "/home/mholub/dronc_seq_temp/input"
output_dir <- "/home/mholub/dronc_seq_temp/output"
```

``` {r inspect}
fastq.files <- list.files(path = data_dir, pattern = ".fastq$", full.names = TRUE)
fastq.files
(cmd <- sprintf("head %s", fastq.files[1]))
system(cmd)
```

```{r QC}
path_to_fastqc <- "/home/mholub/Software/FastQC/fastqc"
system(sprintf("mkdir -p %s/qc", output_dir))

(cmd <- sprintf("%s -o %s/qc %s", 
                path_to_fastqc,
                output_dir,
                fastq.files[1]))
system(cmd)
```

